language/move-compiler/src/naming/ast.rs

```diff
@@ -563,6 +563,80 @@ impl fmt::Display for TypeName_ {
     }
 }
 
+impl fmt::Display for Type_ {
+    fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Type_::Unit => write!(f, "()"),
+            Type_::Ref(mut_, s) => {
+                if *mut_ {
+                    write!(f, "&mut {}", s)
+                } else {
+                    write!(f, "& {}", s)
+                }
+            }
+            Type_::Param(tp) => {
+                if tp.abilities.is_empty() {
+                    write!(f, "{}", tp.user_specified_name)
+                } else {
+                    write!(
+                        f,
+                        "{}: {}",
+                        tp.user_specified_name,
+                        format_delim::<Ability, AbilitySetIntoIter>(
+                            tp.abilities.clone().into_iter(),
+                            " + "
+                        )
+                    )
+                }
+            }
+            Type_::Apply(_, sp!(_, type_name), ss) => match type_name {
+                TypeName_::Multiple(_) => {
+                    write!(f, "({})", format_comma(ss))
+                }
+                TypeName_::Builtin(name) => {
+                    if ss.is_empty() {
+                        write!(f, "{}", name)
+                    } else {
+                        write!(f, "{}<{}>", name, format_comma(ss))
+                    }
+                }
+                TypeName_::ModuleType(sp!(_, module_ident), struct_name) => {
+                    let addr = addr_to_string(&module_ident.address);
+                    if ss.is_empty() {
+                        write!(
+                            f,
+                            "{}::{}::{}",
+                            addr,
+                            module_ident.module.value(),
+                            struct_name
+                        )
+                    } else {
+                        write!(
+                            f,
+                            "{}::{}::{}<{}>",
+                            addr,
+                            module_ident.module.value(),
+                            struct_name,
+                            format_comma(ss)
+                        )
+                    }
+                }
+            },
+            Type_::Var(tv) => write!(f, "var type ({})", tv.0),
```

Comment:
Do you expect this to appear on hover?

Reply:
I do not but since I was writing a generic(ish) implementation of the `Display` trait I wanted to handle it at least semi-gracefully. If we specialize it more towards typ-on-hover, I'd probably just make it into "invalid type".